### PR TITLE
Feature/update es6

### PR DESCRIPTION
**Pull request by akinjide**

### CODE REVIEW

**Comment by andela-abankole** [comment]
@dcodesmith 👀

File: app/controllers/users.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import 'babel-polyfill';
+import db from '../utils/connect';
+import wrap from 'co-monk';
```
Comment (dcodesmith) [inline app/controllers/users.js]: All node modules should be exported first and then you can import local files after, you could also have a line break between them

File: app/controllers/users.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import parse from 'co-body';
+import RequestVerifier from '../utils/request-verifier';
+
+const verify = new RequestVerifier();
+
+console.log(verify.checkRequestBody);
```
Comment (dcodesmith) [inline app/controllers/users.js]: Get rid of all `console.log`

File: app/controllers/users.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * Expose response default.
+ * @param {String} path
+ * @param {String} message
+ * @param {String} status
+ * @return {Object}
+ * @api private
+ */
+function *Handler(path, message, status) {
```
Comment (dcodesmith) [inline app/controllers/users.js]: Not a class or constructor function so should be lowercase

File: app/controllers/users.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import db from '../utils/connect';
+import wrap from 'co-monk';
+import parse from 'co-body';
+import RequestVerifier from '../utils/request-verifier';
+
+const verify = new RequestVerifier();
+
+console.log(verify.checkRequestBody);
+
+const users = wrap(db.get("users"));
```
Comment (dcodesmith) [inline app/controllers/users.js]: Use single quotes everywhere in JS

File: app/controllers/users.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Expose response default.
+ * @param {String} path
+ * @param {String} message
+ * @param {String} status
+ * @return {Object}
+ * @api private
+ */
+function *Handler(path, message, status) {
+  return {
```
Comment (dcodesmith) [inline app/controllers/users.js]: I believe you can do this now ...

```
const current_url = `https://api.degg.com${path}`
return { current_url, message, status }
```

File: app/controllers/users.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    this.throw(403, 'access_denied', error);
+  }
+  else {
+    var insertedUser = yield users.insert(body);
+    this.set('location', "/user/" + insertedUser._id);
+    this.status = 201;
+    this.body = yield* Handler(this.route.path, this.message, this.status);
+  }
+
+  // {
```
Comment (dcodesmith) [inline app/controllers/users.js]: Get rid of all commented code

File: app/controllers/users.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+export function *create(next) {
+  var body = yield parse(this);
+  const error = verify.checkRequestBody(body, ['name', 'city']);
+
+  if (Object.keys(error).length) {
+    this.throw(403, 'access_denied', error);
+  }
+  else {
+    var insertedUser = yield users.insert(body);
+    this.set('location', "/user/" + insertedUser._id);
```
Comment (dcodesmith) [inline app/controllers/users.js]: this.set('location', `/user/${insertedUser._id}`);

File: app/controllers/users.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  return {
+    current_url: `https://api.degg.com${path}`,
+    message: `${message}`,
+    status: status
+  }
+}
+
+
+export function *create(next) {
+  var body = yield parse(this);
```
Comment (dcodesmith) [inline app/controllers/users.js]: Use `const` or `let` instead of `var`

File: app/controllers/users.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  const error = verify.checkRequestBody(body, ['name', 'city']);
+
+  if (Object.keys(error).length) {
```
Comment (dcodesmith) [inline app/controllers/users.js]: You don't need the else, just do ..

```
if (Object.keys(error).length) {
    this.throw(403, 'access_denied', error);
    return;
}
```

Ideally, the next block of code shouldn't execute whenever you `throw` an error. So check it first before you add a return statement.

File: app/controllers/users.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    this.throw(403, 'access_denied', error);
+  }
+  else {
+    var insertedUser = yield users.insert(body);
```
Comment (dcodesmith) [inline app/controllers/users.js]: declare `let insertedUser;` at the top of the function first and then ...

`insertedUser = yield users.insert(body);`

File: app/controllers/users.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    this.set('location', "/user/" + insertedUser._id);
+    this.status = 201;
```
Comment (dcodesmith) [inline app/controllers/users.js]: Consider using [node-http-status](https://github.com/wdavidw/node-http-status) for status codes.

File: app/routes/index.js [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+    current_url: `https://api.degg.com${ctx.route.path}`,
+    message: `degg-api ¯\\_(ツ)_/¯`, 
+    status: ctx.status 
+  };
+
+  return next();
+})
+
+// can use generator middlewares
+router.get('/foobar', function * (next) {
```
Comment (dcodesmith) [inline app/routes/index.js]: What's this for?

File: app/routes/index.js [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  console.log(this.cookies);
```
Comment (dcodesmith) [inline app/routes/index.js]: These logs should get into production. Winston should handle that. Ideally, we shouldn't have then in there.

File: app/routes/users/index.js [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import 'babel-polyfill';
+import Router from 'koa-better-router';
+import { create as addUser, readAll as getUsers } from '../../controllers/users';
```
Comment (dcodesmith) [inline app/routes/users/index.js]: import { create as addUser, readAll as getUsers } from '../../controllers/users';

```
import {
    create as addUser,
    readAll as getUsers
} from '../../controllers/users';
```

```

File: app/utils/connect.js [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+import config from '../../config/config';
+
+/**
+  * connect to mongodb
+  * @public
+  */
+const db = monk(config.mongodb.connectionString);
+db
+  .then((db) => {
+    logger.info(`.`);
```
Comment (dcodesmith) [inline app/utils/connect.js]: What are you logging?

File: app/utils/request-verifier.js [hunk 16]
```diff
@@ -0,0 +1,17 @@
+import _ from 'lodash';
+
+class RequestVerifier {
+  constructor() {
+    this.errors = {}
+  }
+
+  checkRequestBody(params, requiredFields) {
+    for (let i = 0; i < requiredFields.length; i++) {
+      if (!params[requiredFields[i]]) this.errors[`.${requiredFields[i]}`] = 'is required'
```
Comment (dcodesmith) [inline app/utils/request-verifier.js]: Keep it clean and readable.
```
if (!params[requiredFields[i]]) {
    this.errors[`.${requiredFields[i]}`] = 'is required';
}
```

File: bin/www [hunk 17]
```diff
@@ -1,84 +1,85 @@
...
+  }
+
+  const bind = typeof port === 'string'
+    ? `Pipe ${port}`
+    : `Port ${port}`;
+
+  // handle specific listen errors with friendly messages
+  switch (error.code) {
+    case 'EACCES':
+      console.error(`${bind} requires elevated privileges`);
```
Comment (dcodesmith) [inline bin/www]: ?

File: config/logger.js [hunk 18]
```diff
@@ -1,73 +1,83 @@
-const winston = require('winston');
```
Comment (dcodesmith) [inline config/logger.js]: 👍

File: index.js [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+import config from './config/config';
+import fs from 'fs';
+
+import routes from './app/routes';
+
+const app = new koa();
+const api = Router({ prefix: '/v1' });
+
+const logDirectory = __dirname + '/logs';
+fs.existsSync(logDirectory) || fs.mkdirSync(logDirectory);
```
Comment (dcodesmith) [inline index.js]: consider using [fs-extra#ensuredir](https://www.npmjs.com/package/fs-extra#ensuredirdir-callback), it creates the dir for you it doesn't already exist.

File: index.js [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+ * Middlewares.
+ * logger -> `Klogger`
+ * logger -> `morgan`
+ * x-response-time -> `responseTime`
+ * compression -> `compress`
+ * rate limiting -> `ratelimit`
+ * routing -> `api`
+ * @private
+ */
+if ('test' !== app.env) app.use(convert(Klogger()));
```
Comment (dcodesmith) [inline index.js]: Readability ...

```
if ('test' !== app.env) {
    app.use(convert(Klogger()));
}

```

File: package.json [hunk 21]
```diff
@@ -21,20 +21,26 @@
...
+    "koa-morgan": "^1.0.1",
+    "koa-multer": "^1.0.0",
     "koa-ratelimit": "^2.4.0",
     "koa-response-time": "^1.0.2",
-    "koa-route": "^2.4.2",
     "lodash": "^4.16.4",
+    "loggly-winston-bulk": "^1.0.1",
     "monk": "^3.1.3",
     "redis": "^2.6.2",
+    "should": "^11.1.1",
```
Comment (dcodesmith) [inline package.json]: I think `should` should be in `devDependencies`

File: test/root.js [hunk 22]
```diff
@@ -0,0 +1,20 @@
+const request = require('supertest');
+const assert = require('assert');
+
+const app = require('../build/index.js');
+const _request = request.agent(app.default.listen());
+
+describe('# v1/', () => {
+  context('when present', () => {
```
Comment (dcodesmith) [inline test/root.js]: You might also want to test for "when not present"

File: test/users.js [hunk 23]
```diff
@@ -1,22 +1,76 @@
-import co from 'co';
-import users from './userRoutes';
-import app from './index.js';
-import request from 'supertest';
+const co = require('co');
+const db = require('../build/app/utils/connect');
```
Comment (dcodesmith) [inline test/users.js]: move below all node modules and include a line break

File: test/users.js [hunk 24]
```diff
@@ -1,22 +1,76 @@
...
+				.get('/v1/users')
+				.expect(200)
+				.end((err, res) => {
+					res.body.users.length.should.equal(1);
+					res.body.users[0].name.should.eql(seed.name);
+					done();
+				});
+		});
+
+		// it('should get user', done => {
```
Comment (dcodesmith) [inline test/users.js]: ?

File: index.js (hunk 19 above)
Comment (dcodesmith) [inline index.js]: try {
    fs.existsSync(logDirectory) || fs.mkdirSync(logDirectory);
} catch (exception) {
  throw exception;
}

File: index.js [hunk 25]
```diff
@@ -0,0 +1,213 @@
...
+  * koa appication name
+  * @public
+  */
+app.name = 'degg-api';
+
+const logger = app.env
+  ? loggerInit(app.env)
+  : loggerInit();
+
+global.logger = logger;
```
Comment (dcodesmith) [inline index.js]: You shouldn't set globals. http://stackoverflow.com/questions/5447771/node-js-global-variables/9590297#9590297

Comment (akinjide) [inline index.js]: There was a reason for that. I promise

**Pull request merged**